### PR TITLE
feat: Packagesprops support

### LINKED ISSUE #5403: Add support of .NET `*Packages.props` files.
## Description
Add support of [Directory.Packages.props](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management) files.
Also we can parse `Packages.props`  files. These files were used before `Directory.Packages.props` and use same format - https://github.com/aquasecurity/trivy/discussions/5379#discussioncomment-7316578

### PR DESCRIPTION
**Pull request by yuriShafet**
## Description

Adding support and documentation for Directory.packages.props and packages.props.
I am not 100% sure about documentation about transitive dependencies support.
## Related issues
- Close #5403 

## Related PRs
- https://github.com/aquasecurity/go-dep-parser/pull/268

example of work:
```bash
➜  trivy fs ./Directory.Packages.props 
2023-11-22T13:50:17.728+0600    INFO    Vulnerability scanning is enabled
2023-11-22T13:50:17.729+0600    INFO    Secret scanning is enabled
2023-11-22T13:50:17.729+0600    INFO    If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-11-22T13:50:17.729+0600    INFO    Please see also https://aquasecurity.github.io/trivy/dev/docs/scanner/secret/#recommendation for faster secret detection
2023-11-22T13:50:17.739+0600    INFO    Number of language-specific files: 1
2023-11-22T13:50:17.739+0600    INFO    Detecting packages-props vulnerabilities...

Directory.Packages.props (packages-props)

Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 1, CRITICAL: 0)

┌─────────────────┬─────────────────────┬──────────┬────────┬───────────────────┬───────────────┬───────────────────────────────────────────────────┐
│     Library     │    Vulnerability    │ Severity │ Status │ Installed Version │ Fixed Version │                       Title                       │
├─────────────────┼─────────────────────┼──────────┼────────┼───────────────────┼───────────────┼───────────────────────────────────────────────────┤
│ Newtonsoft.Json │ GHSA-5crp-9r3c-p9vr │ HIGH     │ fixed  │ 9.0.1             │ 13.0.1        │ Improper Handling of Exceptional Conditions in    │
│                 │                     │          │        │                   │               │ Newtonsoft.Json                                   │
│                 │                     │          │        │                   │               │ https://github.com/advisories/GHSA-5crp-9r3c-p9vr │
└─────────────────┴─────────────────────┴──────────┴────────┴───────────────────┴───────────────┴───────────────────────────────────────────────────┘
```

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/fanal/types/const.go [hunk 1]
```diff
@@ -95,8 +95,10 @@ const (
 
 // Language-specific file names
 const (
-	NuGetPkgsLock   = "packages.lock.json"
-	NuGetPkgsConfig = "packages.config"
+	NuGetPkgsLock           = "packages.lock.json"
+	NuGetPkgsConfig         = "packages.config"
+	NuGetDirectoryPkgsProps = "directory.packages.props"
+	NuGetPkgsProps          = "packages.props"
```
Comment (DmitriyLewen) [inline pkg/fanal/types/const.go:101]: maybe we can use only `packages.props` for both file names?

File: docs/docs/coverage/language/dotnet.md [hunk 2]
```diff
@@ -11,11 +11,13 @@ The following scanners are supported.
...
-|      NuGet      | packages.config    |            ✓            |     Excluded     |                  -                   |    -     |
-|      NuGet      | packages.lock.json |            ✓            |     Included     |                  ✓                   |    ✓     |
+| Package manager | File                         | Transitive dependencies | Dev dependencies | [Dependency graph][dependency-graph] | Position |
+|:---------------:|------------------------------|:-----------------------:|:----------------:|:------------------------------------:|:--------:|
+|    .Net Core    | *.deps.json                  |            ✓            |     Excluded     |                  -                   |    ✓     |
+|      NuGet      | packages.config              |            ✓            |     Excluded     |                  -                   |    -     |
+|      NuGet      | packages.props               |            -            |     Excluded     |                  -                   |    -     |
+|      NuGet      | Directory.packages.props     |            -            |     Excluded     |                  -                   |    -     |
```
Comment (DmitriyLewen) [inline docs/docs/coverage/language/dotnet.md:19]:
```suggestion
|      NuGet      | *packages.props               |            -            |     Excluded     |                  -                   |    -     |
```

1 line for packages.props will be enough.

But can you add new section (as for `*.deps.json`, `packages.config`, etc.).
In this section you can write about `Directory.Packages.props`, that  `Packages.props` is outdated, etc.

File: pkg/fanal/analyzer/language/dotnet/nuget/nuget.go [hunk 3]
```diff
@@ -8,12 +8,14 @@ import (
...
 	"path/filepath"
 	"sort"
+	"strings"
 
 	"golang.org/x/exp/slices"
 	"golang.org/x/xerrors"
 
 	"github.com/aquasecurity/go-dep-parser/pkg/nuget/config"
 	"github.com/aquasecurity/go-dep-parser/pkg/nuget/lock"
+	"github.com/aquasecurity/go-dep-parser/pkg/nuget/packagesprops"
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/dotnet/nuget/nuget.go:18]: `*Packages.prop` files don't require helping files and are not large in size.
So you can use [analyzer](https://github.com/aquasecurity/trivy/blob/ed0022b91547d582e14fb05731520ea0f61a90d3/pkg/fanal/analyzer/analyzer.go#L71-L76)

**Review by DmitriyLewen — COMMENTED** [review]
Hello @yuriShafet 
Thanks for your work!

Left few comments.

Also 2 moments:
- please take a look at `purl` package. I think you need to add new type.
- can you add small integration test to [repo_test](https://github.com/aquasecurity/trivy/blob/main/integration/repo_test.go)?

**Comment by yuriShafet** [comment]
@DmitriyLewen 
I added packagesprops to integration test. It seems that it is failing at the moment. I compiled trivy locally and executed manual scan of the test repo and got following output:

`
./trivy fs -f json --list-all-pkgs ../../integration/testdata/fixtures/repo/packagesprops/packages.props 

2023-11-21T23:31:21.913-0500	INFO	Vulnerability scanning is enabled

2023-11-21T23:31:21.913-0500	INFO	Secret scanning is enabled

2023-11-21T23:31:21.914-0500	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning

2023-11-21T23:31:21.914-0500	INFO	Please see also https://aquasecurity.github.io/trivy/dev/docs/scanner/secret/#recommendation for faster secret detection

2023-11-21T23:31:21.915-0500	INFO	Number of language-specific files: 0

{

  "SchemaVersion": 2,

  "CreatedAt": "2023-11-21T23:31:21.915709186-05:00",

  "ArtifactName": "../../integration/testdata/fixtures/repo/packagesprops/packages.props",

  "ArtifactType": "filesystem",

  "Metadata": {

    "ImageConfig": {

      "architecture": "",

      "created": "0001-01-01T00:00:00Z",

      "os": "",

      "rootfs": {

        "type": "",

        "diff_ids": null

      },

      "config": {}

    }

  }

}
`

It seems packagesprops logic is not registered somewhere.

File: docs/docs/coverage/language/index.md [hunk 4]
```diff
@@ -34,6 +34,7 @@ On the other hand, when the target is a post-build artifact, like a container im
 | [.NET](dotnet.md)    | packages.lock.json                                                                         |     ✅     |     ✅      |       ✅        |       ✅        |
 |                      | packages.config                                                                            |     ✅     |     ✅      |       ✅        |       ✅        |
 |                      | .deps.json                                                                                 |     ✅     |     ✅      |       ✅        |       ✅        |
+|                      | *Packages.props[^11]                                                                       |     ✅     |     ✅      |       ✅        |       ✅        |
```
Comment (DmitriyLewen) [inline docs/docs/coverage/language/index.md:37]: We divide all language files into “before build” and “after build” - https://aquasecurity.github.io/trivy/v0.47/docs/coverage/language/#supported-languages
Perhaps we should only scan Packages.props files in fs/repo mode?

File: pkg/fanal/analyzer/all/import.go [hunk 5]
```diff
@@ -12,6 +12,7 @@ import (
 	_ "github.com/aquasecurity/trivy/pkg/fanal/analyzer/language/dart/pub"
 	_ "github.com/aquasecurity/trivy/pkg/fanal/analyzer/language/dotnet/deps"
 	_ "github.com/aquasecurity/trivy/pkg/fanal/analyzer/language/dotnet/nuget"
+	_ "github.com/aquasecurity/trivy/pkg/fanal/analyzer/language/dotnet/packagesprops"
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/all/import.go:15]: > It seems packagesprops logic is not registered somewhere.

You needed to add `packagesprops` package here to enable parsing of `*Packages.props` files.

File: pkg/fanal/analyzer/language/dotnet/packagesprops/packagesprops.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	res, err := language.Analyze(types.PackagesProps, input.FilePath, input.Content, parser)
+	if err != nil {
+		return nil, xerrors.Errorf("*Packages.props dependencies analysis error: %w", err)
+	}
+
+	return res, nil
+}
+
+func (a packagesPropsAnalyzer) Required(filePath string, _ os.FileInfo) bool {
+	return strings.HasSuffix(filePath, packagesPropsSuffix)
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/dotnet/packagesprops/packagesprops.go:38]: it looks like `Packages.props` always uses uppercase for first `P` - https://github.com/dotnet/roslyn-tools/blob/b4c5220f5dfc4278847b6d38eff91cc1188f8066/src/RoslynInsertionTool/RoslynInsertionTool/CoreXT.cs#L39-L40
We don't need to check filenames for lower case.

**Review by DmitriyLewen — COMMENTED** [review]
@yuriShafet 
Thanks for your work!

To save us time, I did a little refactoring and added a few comments.
Can you check and confirm that it is working correctly?

File: pkg/fanal/analyzer/language/dotnet/packagesprops/packagesprops.go (hunk 6 above)
Comment (yuriShafet) [reply pkg/fanal/analyzer/language/dotnet/packagesprops/packagesprops.go:38]: I just ran an experiment with `packages.props` file and it seems to work fine

Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/dotnet/packagesprops/packagesprops.go]: Okay, thanks!
I reverted changes in https://github.com/aquasecurity/trivy/pull/5605/commits/c60fdf215c984bfc9d32bfd5e293d93f68faf472

File: docs/docs/coverage/language/index.md (hunk 4 above)
Comment (yuriShafet) [reply docs/docs/coverage/language/index.md:37]: Is there any reason why we should not also support it in Image?

**Comment by yuriShafet** [comment]
@DmitriyLewen  It seems that after your refactor all tests pass. Thanks. Are there any more questions/comments?

File: docs/docs/coverage/language/index.md (hunk 4 above)
Comment (yuriShafet) [reply docs/docs/coverage/language/index.md:37]: By the way, same thing could be said about `packages.lock.json`. I think there will be mostly .dll files in post-build.

Comment (yuriShafet) [reply docs/docs/coverage/language/index.md:37]: Also, If I use my image for development purposes (For example I pre-build an image with the source code to run unit tests or some static code analysis), it is actually make sense to look for these artifacts

Comment (DmitriyLewen) [reply docs/docs/coverage/language/index.md:37]: You are right.
Looks like we need to use pre-build for all supported NuGet files.
But then NuGet packages will not be scanned in `image` mode.

Okay, let's use same logic for `*Packages.props`.

**Review by DmitriyLewen — APPROVED** [review]
@yuriShafet Thanks for your work and help!
I don't have more questions.

@knqyf263 I approved this PR. Take a look, when you have time, please.

**Pull request merged**